spec/lib/s3_config_manager_spec.rb

```diff
@@ -0,0 +1,62 @@
+require 'rails_helper'
+require 's3_config_manager'
+
+RSpec.describe S3ConfigManager do
+  let(:bucket) { 'example-prod-secrets' }
+  let(:env_name_file) { Tempfile.new('test') }
+  let(:env_name) { 'staging' }
+
+  before { File.open(env_name_file.path, 'w') { |file| file.puts(env_name) } }
+  after { env_name_file.unlink }
+
+  let(:s3_client) { FakeS3.new }
+  subject(:s3_config_manager) do
+    S3ConfigManager.new(
+      bucket: bucket,
+      env_name_path: env_name_file.path,
+      s3_client: s3_client
+    )
+  end
+
+  describe '.download_configs' do
+    let(:local_config_file) { Tempfile.new('test') }
+
+    let(:config_files) do
+      { '/v1/idp/%{env_name}/some_config.yml' => local_config_file.path }
+    end
+
+    after { local_config_file.unlink }
+
+    subject(:download_configs) { s3_config_manager.download_configs(config_files) }
+
+    context 'when the env file does not exist' do
+      before { FileUtils.rm(env_name_file.path) }
+
+      it 'prints an error and does not download anything from s3' do
+        expect($stderr).to receive(:puts).
+          with("S3ConfigManager: no file exists at #{env_name_file.path}")
+        expect(s3_client).to_not receive(:get_object)
+
+        download_configs
+      end
+    end
+
+    context 'when the env file exists' do
+      let(:config_body) { 'test config data' }
+
+      before do
+        s3_client.put_object(
+          bucket: bucket,
+          key: "v1/idp/#{env_name}/some_config.yml",
```

Comment:
`env_name` should come first, so we have a way to control permissions with IAM per environment.